code/modules/urist/modules/shipbattles/overmapships/components/advanced_components.dm

```diff
@@ -0,0 +1,322 @@
+//boarding
+/datum/shipcomponents/teleporter
+	name = "teleporter module"
+	var/boarding_delay = 3 MINUTES //How long between boarding attempts?
+	var/list/boarding_mobs = list(/mob/living/simple_animal/hostile/russian, /mob/living/simple_animal/hostile/russian/ranged) //What mobs do we spawn on boarding?
+	var/boarding_number = 4 //We'll spawn this many every time we board.
+	var/last_boarding = 0 //How long since we boarded last?
+	var/boarding_turf = null//Where do we spawn our mobs?
+	var/boarding_failure_chance = 0 //more shields = more chance of preventing teleportation.
+	var/boarded_max = 3 //how many times do we board?
+	var/boarded_amount = 0 //how many times have we boarded
+	var/boarding_message = "Enemy teleporter locked on! Boarding imminient! Expected breach point:"
+	var/initial_delay = FALSE
+	var/bypass_shields = FALSE
+	var/counterboarding = FALSE
+
+/datum/shipcomponents/teleporter/DoActivate()
+	if(mastership.boarding && !counterboarding)
+		return //they have better things to do than board the Nerva right now.
+
+	if(initial_delay)
+		mastership.target_ship.autoannounce("<b>Hostile [name] detected powering up. Expected time until ready: [(boarding_delay/600)] minutes.</b>", "public")
+		last_boarding = world.time + boarding_delay
+		return
+
+	if(last_boarding > world.time) //Are we too early?
+		return
+
+	if(boarded_amount >= boarded_max)
+		return
+
+	boarding_failure_chance = 0 //Zero this var out.
+
+	if(!bypass_shields)
+		for(var/obj/machinery/power/shield_generator/S in SSmachines.machinery) //Calculate our failure chance.
+			if(S.z in mastership.target_ship.map_z)
+				if(S.running == SHIELD_RUNNING)
+					boarding_failure_chance += 25 // four shield generators to TOTALLY block boarding.
+
+	if(!boarding_turf) //Locate where we're boarding, give them a warning.
+		var/obj/item/device/radio/beacon/active_beacon //what beacon are we locking onto?
+		var/list/beacon_list = list()
+		for(var/obj/item/device/radio/beacon/B in GLOB.listening_objects)
+			if(B.z in mastership.target_ship.map_z)
+				beacon_list += B
```

Comment:
This is no good if we have to calculate this every MC tick(). Need to consider moving this off to somewhere else.